### PR TITLE
[IAP]Refactor the in-app purchase of Google Play

### PR DESCRIPTION
**Pull request by minggangw**
This patch refactors the previous implementation of in-app purchase(IAP)
on Android. The main differences are:

1. Redefine the javascript interfaces to combine with the one used on
iOS, thus developers can deploy the same code on both platforms. The
spec locates at https://github.com/crosswalk-project/ios-extensions-crosswalk
/blob/master/extensions/InAppPurchase/spec/iap.html.

2. Refactor the framework of the IAP extension in order to add
additional stores later. The extension itself acts as a manager of
different backends, it forwards the commands to the proper backend if it
supports. To add support of new backend, just inherit the
class InAppPurchaseHelper and implement the command methods.

3. Refactor the implementation of Google Play to reject the promise when
an error happens.

BUG=XWALK-5696

### CODE REVIEW

**Comment by minggangw** [comment]
@halton @jondong Please review this PR, thanks!

File: build/common.gypi [hunk 1]
```diff
@@ -11,5 +11,6 @@
     'android_sdk%': '<!(ls -d <(android_sdk_root)/platforms/*|tail -1)',
     'android_sdk_tools%': '<!(dirname $(which aidl))',
     'app_runtime_java_jar%': '<!(find -L $(dirname $(which make_apk.py)) -name xwalk_app_runtime_java.jar|tail -1)',
+    'core_library_java_jar%': '<!(find -L $(dirname $(which make_apk.py)) -name xwalk_core_library_java.jar|tail -1)',
```
Comment (halton) [inline build/common.gypi]: is `app_runtime_java_jar` still needed?

File: iap/README [hunk 2]
```diff
@@ -1,4 +0,0 @@
-How to package the IAP extension with your application:
-
-  cd $(dirname $(which make_apk.py))
-  python make_apk.py --name=iap --package=org.xwalk.extensions.iap --permissions=iap --extensions="/<path-to-xwalk-android-extension>/out/Default/gen/iap/" --arch=x86 --manifest=/<path-to-xwalk-android-extension>/examples/iap/iap.json
```
Comment (halton) [inline iap/README]: You suppose have an updated README instead of removing it.

File: iap/iap.js [hunk 3]
```diff
@@ -1,59 +1,113 @@
-// Copyright (c) 2014 Intel Corporation. All rights reserved.
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
```
Comment (halton) [inline iap/iap.js]: Please leave 2014 copyright, you only need to have new year copyright for new files.

File: iap/iap.js [hunk 4]
```diff
@@ -1,59 +1,113 @@
...
+
+    if (request) {
+        if (msg.error) {
+            request.reject.apply(null, [msg.error]);
+        } else {
+            request.resolve.apply(null, [msg.data]);
+        }
+        delete g_asyncRequests[msg.requestId];
+    }
+});
```
Comment (halton) [inline iap/iap.js]: Please follow google javascript style, 2-space indent is one of I noticed.

File: iap/src/org/xwalk/extensions/InAppPurchaseExtension.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        public Integer requestId;
+        public InAppPurchaseHelper helper;
+    }
+
+    private InAppPurchaseHelper createHelperByChannel(int instanceId, String channel) {
+        if (channel.equals(BACKEND_GOOGLE)) {
+            return new InAppPurchaseGoogleHelper(mExtensionContext.getActivity(), instanceId, mHelperListener);
+        } else if (channel.equals(BACKEND_XIAOMI)) {
+            // TODO(minggang): add the support of XiaoMi, when it's ready.
+        }
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseExtension.java]: I think need a message to indicate unsupported store name.

File: iap/src/org/xwalk/extensions/InAppPurchaseExtension.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        return null;
+    }
+
+    private InAppPurchaseHelper.InAppPurchaseHelperListener mHelperListener =
+        new InAppPurchaseHelper.InAppPurchaseHelperListener() {
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseExtension.java]: wrong indent, 8-space for unfinished line

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        mIab.dispose();
+        mIab = null;
+      }
+    }
+
+    @Override
+    public void handleActivityResult(int requestCode, int resultCode, Intent data) {
+        if (mIab != null) {
+            mIab.handleActivityResult(requestCode, resultCode, data);
+        }
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: `if (mIab != null) mIab.handleActivityResult(requestCode, resultCode, data);`

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    // Callback for when a query for product details is finished.
+    private IabHelper.QueryInventoryFinishedListener mGotDetailsListener =
+        new IabHelper.QueryInventoryFinishedListener() {
+            public void onQueryInventoryFinished(IabResult result, Inventory inventory) {
+                if (mIab == null)
+                    return;
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: squash into one line.

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    @Override
+    public void handleActivityResult(int requestCode, int resultCode, Intent data) {
+        if (mIab != null) {
+            mIab.handleActivityResult(requestCode, resultCode, data);
+        }
+    }
+
+    // Callback for when a query for product details is finished.
+    private IabHelper.QueryInventoryFinishedListener mGotDetailsListener =
+        new IabHelper.QueryInventoryFinishedListener() {
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: indent

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+                        List<SkuDetails> skuList = inventory.getAllProducts();
+                        JSONArray jsonSkuList = new JSONArray();
+                        for (SkuDetails sku : skuList) {
+                            jsonSkuList.put(sku.toJson());
+                        }
+                        mListener.onRequestFinished(mInstanceId, jsonSkuList);
+                    } catch (JSONException e) {
+                        Log.e(TAG, e.toString());
+                    }
+                }
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: Prefer
```
if (result.isFailure()) {
    mListener.onRequestFailed(mInstanceId, OPERATION_ERROR);
    return;
}

try {
...
```

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        };
+
+    // Callback for when a purchase is finished.
+    private IabHelper.OnIabPurchaseFinishedListener mPurchaseFinishedListener =
+        new IabHelper.OnIabPurchaseFinishedListener() {
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: indent

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+            public void onIabPurchaseFinished(IabResult result, Purchase purchase) {
+                if (mIab == null)
+                    return;
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: ditto

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+                        public void run() {
+                            try {
+                                mIab.flagEndAsync();
+                                mIab.consumeAsync(p, mConsumeFinishedListener);
+                            } catch (java.lang.IllegalStateException e) {
+                                Log.e(TAG, e.toString());
+                            }
+                        }
+                    });
+                }
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: ditto

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        };
+
+
+    // Called when consumption is completed.
+    IabHelper.OnConsumeFinishedListener mConsumeFinishedListener =
+        new IabHelper.OnConsumeFinishedListener() {
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: indent

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+            public void onConsumeFinished(Purchase purchase, IabResult result) {
+                if (mIab == null)
+                    return;
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: ditto

File: iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+                            purchaseResult.put("packageName", purchase.getPackageName());
+                            purchaseResult.put("productId", purchase.getSku());
+                            purchaseResult.put("purchaseTime", purchase.getPurchaseTime());
+                            mListener.onRequestFinished(mInstanceId, purchaseResult);
+                        } catch (JSONException e) {
+                            Log.e(TAG, e.toString());
+                        }
+                    } else {
+                        mListener.onRequestFailed(mInstanceId, OPERATION_ERROR);
+                    }
```
Comment (halton) [inline iap/src/org/xwalk/extensions/InAppPurchaseGoogleHelper.java]: ditto

File: iap/README (hunk 2 above)
Comment (minggangw) [inline iap/README]: This content of the README should be for the sample project, not the extension, so I removed it and will add to the sample.

File: build/common.gypi (hunk 1 above)
Comment (minggangw) [inline build/common.gypi]: Other extensions still need this, will delete when the extensions get updated.

**Comment by minggangw** [comment]
PR updated, please have a look.

**Comment by minggangw** [comment]
PR updated,  @jondong @halton pls review, thanks!

**Comment by minggangw** [comment]
Thanks @deqing , nice to meet you! :)

**Comment by deqing** [comment]
Thanks @minggangw, the PR look good to me. It's nice to have IAP back to work.

**Pull request merged**